packages/server/src/notifier/model/email/utils/index.ts

```diff
@@ -1,16 +1,62 @@
 import { Prisma } from '@prisma/client'
+import { request } from 'graphql-request'
 
-import { Email, EmailBody } from '@/common/utils/email'
+import { QUERY_NODE_ENDPOINT } from '@/common/config'
+import { GetPostDocument, GetThreadDocument } from '@/common/queries'
+import { Email } from '@/common/utils/email'
 
 export type Notification = Prisma.NotificationGetPayload<{ include: { member: true } }>
 type NotificationWithEmailAddress = Notification & { member: { email: string } }
 
-type PartialEmail = Pick<Email, 'subject'> & EmailBody
-export type EmailFromNotification = (notification: NotificationWithEmailAddress) => Promise<Email> | undefined
-
-export const buildEmail =
-  <T>(to: string, getData: () => Promise<T>) =>
-  async (fromData: (data: T) => PartialEmail): Promise<Email> => ({ ...fromData(await getData()), to })
+export type EmailFromNotificationFn = (notification: NotificationWithEmailAddress) => Promise<Email | void>
 
 export const hasEmailAddress = (notification: Notification): notification is NotificationWithEmailAddress =>
   !!notification.member.email
+
+interface ForumPost {
+  author: string
+  threadId: string
+  thread: string
+  text: string
+}
+const cachedForumPosts: { [id: string]: ForumPost } = {}
+export const getForumPost = async (id: string): Promise<ForumPost> => {
+  if (!(id in cachedForumPosts)) {
+    const { forumPostByUniqueInput: post } = await request(QUERY_NODE_ENDPOINT, GetPostDocument, { id })
+    if (!post) {
+      throw Error(`Failed to fetch post ${id} on the QN`)
+    }
+
+    cachedForumPosts[id] = {
+      author: post.author.handle,
+      threadId: post.thread.id,
+      thread: post.thread.title,
+      text: post.text,
+    }
+  }
+
+  return cachedForumPosts[id]
+}
+
+interface ForumThread {
+  author: string
+  title: string
+  text?: string
+}
+const cachedForumThreads: { [id: string]: ForumThread } = {}
+export const getForumThread = async (id: string): Promise<ForumThread> => {
+  if (!(id in cachedForumThreads)) {
+    const { forumThreadByUniqueInput: thread } = await request(QUERY_NODE_ENDPOINT, GetThreadDocument, { id })
+    if (!thread) {
+      throw Error(`Failed to fetch thread ${id} on the QN`)
+    }
+
+    cachedForumThreads[id] = {
+      author: thread.author.handle,
+      title: thread.title,
+      text: thread.initialPost?.text,
+    }
+  }
+
+  return cachedForumThreads[id]
+}
```

Comment:
Since there will be a lot of these queries based on entity ids I think it would be nice to move these two functions to a new `./forum.ts` file (in this folder) exported by this `./index.ts`

Reply:
moved 👍